### PR TITLE
Reuse MappingMetadata instances in Metadata class.

### PR DESCRIPTION
**Pull request by martijnvg**
Hash the mapping source of a MappingMetadata instance and then
cache it in Metadata class. A mapping with the same hash
will use a cached MappingMetadata instance. This can
significantly reduce the number of MappingMetadata instances
for data streams and index patterns.

Idea originated from #69772, but just focusses on the jvm heap memory savings.
And hashes the mapping instead of assigning it an uuid.

Relates to #77466

### CODE REVIEW

**Comment by martijnvg** [comment]
Locally I have been testing this improvement with the script at the end of this comment. This script creates a data stream and rolls it over concurrently (1000 rollovers). With this change, the number of `MappingMetadata` instances is 3 (running the script in en empty cluster) and without this change over 2000 instances.

```python
import threading
from elasticsearch import Elasticsearch

ds_base_name = 'logs-mysql-'
es = Elasticsearch(['http://localhost:9200'], http_auth=('elastic-admin', 'elastic-password'), timeout=600)

def rollover_test(ds_name):
    es.indices.delete_data_stream(name=ds_name, ignore=[400, 404])
    es.indices.create_data_stream(name=ds_name)
    for i in range(200):
        es.indices.rollover(alias=ds_name,wait_for_active_shards=0)


if __name__ == "__main__":
    es.cluster.put_settings(body={"persistent": {"cluster.max_shards_per_node": 100000}})
    es.indices.put_index_template(name='my-template', body={
        "index_patterns": ["logs-*-*"],
        "priority": 200,
        "data_stream": {},
        "composed_of": [
            "logs-mappings",
            "data-streams-mappings"
        ],
        "template": {
            "settings": {
                "index.number_of_replicas": 0
            }
        },
        "allow_auto_create": True
    })

    for i in range(5):
        t = threading.Thread(target=rollover_test, args=(ds_base_name + str(i),))
        t.start()
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 1]
```diff
@@ -1911,6 +1935,81 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
...
+            if (entry != null) {
+                cache.put(digest, new Entry(entry.refCounter + 1, entry.mapping));
+                indexMetadataBuilder.putMapping(entry.mapping);
+            } else {
+                entry = new Entry(1, indexMetadataBuilder.mapping());
+                cache.put(digest, entry);
+            }
+        }
+
+        private void cleanupUnusedEntry(IndexMetadata previous) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1974]: Because the cache is passed down to new `Metadata` instances, a cleanup mechanism is needed.
This approach with counting the times a `MappignMetata` instance is used and removing entries when there are no usages seems cheaper compared to checking each metadata is build whether all entries are used and then removed the unused entries. But this approach is also more complex. I'm also okay to change the cleanup logic to be simpler and checking each cache entry whether it is used in any of the indices and if unused remove.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java [hunk 2]
```diff
@@ -41,36 +45,53 @@
...
+        this.digest = computeDigest(mapping);
     }
 
-    @SuppressWarnings("unchecked")
     public MappingMetadata(String type, Map<String, Object> mapping) {
         this.type = type;
+        String mappingAsString = Strings.toString((builder, params) -> builder.mapContents(mapping));
         try {
-            this.source = new CompressedXContent((builder, params) -> builder.mapContents(mapping));
+            this.source = new CompressedXContent(mappingAsString);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:72]: Can't we use `this.source = new CompressedXContent((builder, params) -> builder.mapContents(mapping));` here still and compute the digest as we compress for example since we stream the bytes in that case anyway?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java [hunk 3]
```diff
@@ -88,13 +109,12 @@ public static void writeMappingMetadata(StreamOutput out, ImmutableOpenMap<Strin
...
 
-    @SuppressWarnings("unchecked")
-    private boolean routingRequired(Map<String, Object> withoutType) {
+    private boolean routingRequired(Map<?, ?> withoutType) {
         boolean required = false;
         if (withoutType.containsKey("_routing")) {
-            Map<String, Object> routingNode = (Map<String, Object>) withoutType.get("_routing");
-            for (Map.Entry<String, Object> entry : routingNode.entrySet()) {
-                String fieldName = entry.getKey();
+            Map<?, ?> routingNode = (Map<?, ?>) withoutType.get("_routing");
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:115]: Why can't we have `String` keys here anymore?

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1974]: If the simpler version doesn't come with problematic runtime overhead it might be fine? It seems like it should be cheap now that we can cheaply compare mappings?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:72]: On that note, I wonder if we should just make the `Sha256` value a part of the `CompressedXContent`? I see how we do some magic here to make sure we build a deterministic version based on a map, but we could have that by just adding a static constructor to this thing? For BwC we can compute the CRC or SHA1 dynamically where/if needed I guess.
Also, that would allow us to make the `equals` in `CompressedXContent` almost free by just turning it into a string comparison :) WDYT?

**Review by original-brownbear — COMMENTED** [review]
Thanks Martijn, this looks really nice and should be pretty impactful. I just have a couple of questions, suggestions on the plumbing around this mostly :) Happy to discuss this on another channel if that's quicker as well.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 1 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1974]: I think that is true. Also I find this implementation a bit fragile. I will change the cleanup implementation to just check for each cache entry whether it is used by any IndexMetadata.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java (hunk 3 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:115]: I was a bit annoyed with the `SuppressWarnings("unchecked")` and so removed it. I can undo this change and make this change in another pr.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java (hunk 2 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:72]: I like this idea and in fact I did think about this while working on this. I will try this out and see whether it is not too difficult.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 1 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: Implemented: 15b05afb51274069320f1db6ae29f62b95e91348

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java (hunk 2 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java]: I did the first part of this: c7705231c9285e75c809619bb80bc0176c131657
I will work on making the sha256 really part of of `CompressedXContent`, just like crc32 checksum.
Just to double check, do we want to keep the `crc32` checksum? I think so.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java]: > Just to double check, do we want to keep the crc32 checksum? I think so.

Not sure tbh. It seems we're only using it to speed up comparisons and as a `hashCode`. Both of which we can have from the sha256 value can't we?

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java]: Right, I thought there was another usage of the crc32, but it only used for hashcode and checking consistency in assertions. Both can be done with sha256 too. I will replace crc32 completely with sha256.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java]: I pushed: 1961453aa16caa9468f541b975a047a15ae56e01

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:92]: Maybe, this one is a little too impactful to have additional noise in it I think :)

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 4]
```diff
@@ -230,7 +231,8 @@ private Metadata(
         String[] visibleOpenIndices,
         String[] allClosedIndices,
         String[] visibleClosedIndices,
-        SortedMap<String, IndexAbstraction> indicesLookup
+        SortedMap<String, IndexAbstraction> indicesLookup,
+        Map<String, MappingMetadata> cache
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:235]: NIT: Maybe call this something mapping specific, like `mappingsByHash` or so? :) `cache` seems a little generic :)

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 5]
```diff
@@ -1911,6 +1930,60 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
...
+            if (indexMetadata.mapping() == null) {
+                return indexMetadata;
+            }
+
+            String digest = indexMetadata.mapping().getSha256();
+            MappingMetadata entry = cache.get(digest);
+            if (entry != null) {
+                IndexMetadata.Builder imBuilder = new IndexMetadata.Builder(indexMetadata);
+                imBuilder.putMapping(entry);
+                return imBuilder.build();
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1944]: This builder is fairly expensive if I remember correctly with all the setting lookup and the like. I wonder if we could just make this a copy constructor that swaps out the mapping metadata only like we did elsewhere and ourselves the trouble of setting up a builder.
Also, in this case you could make it transparent easily, and not create a new instance even in case the mappings are already instance equal?

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 6]
```diff
@@ -1911,6 +1930,60 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
...
+            if (entry != null) {
+                indexMetadataBuilder.putMapping(entry);
+            } else {
+                cache.put(digest, indexMetadataBuilder.mapping());
+            }
+        }
+
+        private void purgeMappingCache(ImmutableOpenMap<String, IndexMetadata> indices) {
+            final var iterator = cache.entrySet().iterator();
+            while (iterator.hasNext()) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1967]: This seems needlessly hard :) Can't we just collect all the hash strings from the metadata into a set in one loop and then all keys from `cache` that aren't in that set? That way we're at least O(N)? :)

File: server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java [hunk 7]
```diff
@@ -41,43 +43,43 @@
...
+        MessageDigest messageDigest = MessageDigests.sha256();
         try {
-            data.writeTo(new CheckedOutputStream(Streams.NULL_OUTPUT_STREAM, crc32));
+            data.writeTo(new DigestOutputStream(Streams.NULL_OUTPUT_STREAM, messageDigest));
         } catch (IOException bogus) {
             // cannot happen
             throw new Error(bogus);
         }
-        return (int) crc32.getValue();
+        return MessageDigests.toHexString(messageDigest.digest());
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java:54]: Maybe we should base64 encode these instead to save some bytes?

**Review by original-brownbear — COMMENTED** [review]
Thanks Martijn, this looks really nice (and surprisingly simple as well :)).
Only a couple comments on the details that I missed on the first round, other than that, this looks good to go.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 8]
```diff
@@ -1911,6 +1930,60 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
             XContentParserUtils.ensureExpectedToken(XContentParser.Token.END_OBJECT, parser.nextToken(), parser);
             return builder.build();
         }
+
+        private IndexMetadata reuseMappings(IndexMetadata indexMetadata) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1934]: One more thing, `cache` (or whatever we might rename it to) should have a little bit of JavaDoc that roughly explains what's going on here.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java (hunk 3 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:92]: Ok, I will undo this bit here in the pr.

File: server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java [hunk 9]
```diff
@@ -43,43 +43,43 @@
     private static final ThreadLocal<InflaterAndBuffer> inflater1 = ThreadLocal.withInitial(InflaterAndBuffer::new);
     private static final ThreadLocal<InflaterAndBuffer> inflater2 = ThreadLocal.withInitial(InflaterAndBuffer::new);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java:44]: @martijnvg I just realized, we can probably remove all of this infrastructure here now that we have efficient `equals`? I added this only to speed up comparison a while back, but now it seems irrelevant and can be simplified?

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 4 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: I've renamed this: 3bd8a655ddc506f88297f79d9339138cd198f73b

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 5 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: Something like this: 4bc2229d6139e7c1e2da1a9b830ea8fc7988e63d ?
I think we need to check whether the instances are not the same and I think here instances are always different?

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 6 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1982]: Something like this: 3d2bf906b839b22330b047cc0441e4b3dfb0e040 ?

File: server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java (hunk 7 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java]: I've changed this to use base64: 17f00d92ee1dc481579188491ff8e53c8c7bd733

File: server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java (hunk 9 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java]: Yes, I've removed inflater2 and equalsWhenUncompressed() method: c05791c781e3ca27a56a190bb265a49770da8db3

`inflater1 ` and a few other things are still in use by production code.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 8 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: This was done as part of the rename commit: 3bd8a655ddc506f88297f79d9339138cd198f73b

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java (hunk 3 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java]: Removed this change: de25c023076ed9d533cd785f60bdeaad3bbd4368

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java [hunk 10]
```diff
@@ -561,6 +561,44 @@ private IndexMetadata(
         assert numberOfShards * routingFactor == routingNumShards : routingNumShards + " must be a multiple of " + numberOfShards;
     }
 
+    IndexMetadata(IndexMetadata indexMetadata, MappingMetadata mapping) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:564]: Could we maybe make this just a method:

`IndexMetadata withMappingMetadata(MappingMetadata mapping)` to give us an obvious copy constructor. I also wonder, even though it's slightly more code, maybe it's better if this would just invoke the existing constructor so we don't have two constructors to maintain for this thing (given how additional fields are incoming from settings moving to fields in this class)?

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 6 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1967]: ++ perfect

**Review by original-brownbear — APPROVED** [review]
LGTM, I think I'd like the index-metadata constructor to be set up differently just to enable follow-ups but it's your call, I can deal with that myself in a follow-up when I work on it again :)
Thanks so much for the iterations on this one!

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java (hunk 10 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java]: 👍 I agree this is cleaner. I've made this change: 7d092821070b0983e4280a07be21251647e40636
